lib/upipe-modules/upipe_blit.c

```diff
@@ -218,6 +246,22 @@ static void upipe_blit_sub_input(struct upipe *upipe, struct uref *uref,
 
     ubuf_free(sub->ubuf);
     sub->ubuf = uref_detach_ubuf(uref);
+
+    /* Read subtitle pts and duration */
+
+    int type;
+    uint64_t date, duration;
+    if (unlikely(!ubase_check(uref_clock_get_duration(uref, &duration))))
+        duration = 0;
+
+	uref_clock_get_date_prog(uref, &sub->date_prog, &type);
+	if (type == UREF_DATE_NONE) {
+        sub->date_prog = UINT64_MAX;
+        sub->end_date = UINT64_MAX;
+    } else {
+        sub->end_date = sub->date_prog + duration;
+    }
+
```

Comment:
I'd prefer to keep the old behaviour (subpics are blitted until they are replaced) because some people rely on it (upipe_blit is used for other things than subtitles). You could instead have a pipe in front of upipe_blit that checks duration and emits an empty uref when the subtitle is over. BTW, in upipe_blit there is no check of the input PTS to compare it with the subpic PTS. So to have precise subtitling you need to rework your architecture. We can discuss that on IRC.